### PR TITLE
HTTPClient Enhancements

### PR DESCRIPTION
**Pull request by mattt**
A couple of improvements to `HTTPClient` in support of the Swift package registry implementation (#3023). 

### Motivation:

I want to use `URLProtocol` to test the HTTP client that interacts against a mock package registry server. However, it's not currently possible to access the underlying `URLSession` of an `HTTPClient` to configure the URL protocol. 

### Modifications:

This PR introduces a `HTTPClientProtocol` protocol that provides the minimal amount of dependency injection necessary to make this work.

In addition, this PR conforms `HTTPClientHeaders` to be `ExpressibleByDictionaryLiteral` and reorganizes some declarations in `HTTPClient.swift`. These changes (8385b794c44b1cedde4d778c504ab4a65ef35087 and c5717b04167b6e27585c5ffe21ec009b5af032ab) are entirely severable, and you're welcome to ignore them. 

### Result:

As a result of merging this PR, we'll be able to test `HTTPClient` instances using mocked responses from custom `URLProtocol` subclasses.

### CODE REVIEW

**Comment by mattt** [comment]
@swift-ci please smoke test

**Comment by mattt** [comment]
CI appears to be failing for unrelated reasons. Any idea what's causing this?

<details>

### macOS

/Users/buildnode/jenkins/workspace/swift-package-manager-with-xcode-self-hosted-PR-osx/branch-main/swiftpm/.build/checkouts/swift-driver/Sources/SwiftDriver/Execution/DriverExecutor.swift:91: error: -[BuildTests.BuildPlanTests testExplicitSwiftPackageBuild] : failed: caught error: "jobFailedWithNonzeroExitCode(1, "<unknown>:0: error: unknown argument: \'-scan-dependencies\'\n<unknown>:0: error: unknown argument: \'-placeholder-dependency-module-map-file\'\n")"


#### Linux

10:17:39 
<EXPR>:0: error: BuildPlanTests.testExplicitSwiftPackageBuild : threw error "jobFailedWithNonzeroExitCode(1, "<unknown>:0: error: unsupported target architecture: \'\'\n<unknown>:0: error: unsupported target OS: \'\'\n")"

</details>

**Review by tomerd — APPROVED** [review]
lgtm once CI is green

**Comment by tomerd** [comment]
@owenv @DougGregor these failures seems to be driver related. was "main" branch changes with breaking changes?

**Comment by mattt** [comment]
Possibly related to https://github.com/apple/swift-driver/pull/403, which is now merged. Retrying.

@swift-ci please smoke test

**Comment by mattt** [comment]
@swift-ci please smoke test

**Comment by mattt** [comment]
@swift-ci please smoke test

**Comment by tomerd** [comment]
@mattt fwiw, the way we test in other cases is by injecting a mock handler - see for example https://github.com/apple/swift-package-manager/blob/main/Tests/PackageCollectionsTests/GitHubPackageMetadataProviderTests.swift

**Review by abertelrud — APPROVED** [review]

**Comment by mattt** [comment]
@tomerd CI is ✅, so feel free to merge whenever you're ready.

Thanks for pointing out the use of handlers for testing. If you'd prefer, I can rewrite #3023 to use that instead of `URLProtocol`.

**Comment by tomerd** [comment]
> Thanks for pointing out the use of handlers for testing. If you'd prefer, I can rewrite #3023 to use that instead of URLProtocol.

your call

**Pull request merged**
